File: apps/api/src/modules/runs/runDetail.ts
import type {
  AgentOutput,
  AiUsage,
  FileDiff,
  GitInfo,
  TestReport,
} from '@cpwork/shared';
import { runsRepo } from '../../db/repositories/runs.js';

import type { FileBackup } from '../git/git.service.js';

export interface StoredDetail {
  output: AgentOutput | null;
  diffs: FileDiff[];
  applied: boolean;
  test: TestReport | null;
  git: GitInfo | null;
  usage: AiUsage | null;
  /** Pre-apply snapshots used to revert working-tree changes. */
  backups: FileBackup[];
  planFilePath?: string | null;
}

const EMPTY: StoredDetail = {
  output: null,
  diffs: [],
  applied: false,
  test: null,
  git: null,
  usage: null,
  backups: [],
};

export function loadDetail(runId: string): StoredDetail {
  const json = runsRepo.getDetailJson(runId);
  if (!json) return { ...EMPTY };
  try {
    return { ...EMPTY, ...JSON.parse(json) };
  } catch {
    return { ...EMPTY };
  }
}

export function saveDetail(runId: string, detail: StoredDetail): void {
  runsRepo.setDetailJson(runId, JSON.stringify(detail));
}

export function patchDetail(runId: string, patch: Partial<StoredDetail>): StoredDetail {
  const current = loadDetail(runId);
  const next = { ...current, ...patch };
  saveDetail(runId, next);
  return next;
}

File: apps/api_py/services/run_detail.py
import json
from db.runs import runs_repo

EMPTY_DETAIL = {
    "output": None,
    "diffs": [],
    "applied": False,
    "test": None,
    "deploy": None,
    "git": None,
    "usage": None,
    "backups": [],
    "currentStep": "select",
    "completedSteps": [],
    "jiraSnapshot": None,
    "customTitle": None,
    "customTaskKey": None,
    "customRequirements": None,
    "planMarkdown": None,
    "planFilePath": None,
    "planApprovedAt": None,
    "planApprovedBy": None,
    "approvalStatus": "draft",
    "jiraCommentPostedAt": None,
    "jiraCommentId": None,
    "jiraCommentText": None,
    "testPassRate": None,
}


def load_detail(run_id: str) -> dict:
    json_str = runs_repo.get_detail_json(run_id)
    if not json_str:
        return dict(EMPTY_DETAIL)
    try:
        stored = json.loads(json_str)
        merged = {**EMPTY_DETAIL, **stored}
        if merged.get("diffs") is None:
            merged["diffs"] = []
        if merged.get("completedSteps") is None:
            merged["completedSteps"] = []
        if merged.get("output") and isinstance(merged["output"], dict):
            out = merged["output"]
            if out.get("files") is None:
                out["files"] = []
            if out.get("manualTestChecklist") is None:
                out["manualTestChecklist"] = []
            if out.get("risks") is None:
                out["risks"] = []
        return merged
    except Exception:
        return dict(EMPTY_DETAIL)


def save_detail(run_id: str, detail: dict) -> None:
    runs_repo.set_detail_json(run_id, json.dumps(detail))


def patch_detail(run_id: str, patch: dict) -> dict:
    current = load_detail(run_id)
    next_detail = {**current, **patch}
    save_detail(run_id, next_detail)
    return next_detail
